[FEATURE] Implement DSDHacked specification and initial ID24 support for object lookup

cmbernard333 opened the pull request:
## Summary

Implementation of dsdhacked specification #902 and partial implementation of id24 specification for object lookup.

## Contributions and Callouts

This is the culmination of all the work required to make dsdhacked worked in Odamex.

It would not have been at all possible without the tireless efforts and contributions of @electricbrass.

Thanks also to @Hekksy who tested at every major milestone.

## Specifications

- Implementation of [DSDHacked](https://github.com/kraflab/dsda-doom/blob/master/docs/dsdhacked.md)
- Partial Implementation of [id24](https://docs.google.com/document/d/19uWD7psvohLwQbzYl9Ja6q6SLCT5crN5HPFhUA9lcv8/edit?tab=t.0#heading=h.6jl50zhuy37m) with lookup tables for map objects as well as spawn map support.
- Odamex objects now use id24 negative indices range.

## Misc

- A_LineEffect is now functional fixing bugs in Eviternity I and Eviternity II

## DSDHacked Implementation Details

### Doom Objects

Odamex, being a descendant of ZDoom, had fixed data structures for all "Doom Objects" (Mobj, State, Sprite Names, etc). All of these Doom Objects are now dynamically allocated at runtime with the original fixed data structures used as the base. These are all allocated within the memory zone that is created at startup.

### The `DoomObjectContainer` Class

DoomObjectContainer is the primary data structure for storing and retrieving the doom objects. It is a templated data structure that stores pointers to the objects that are allocated in the memory zone. It is structured internally with a `std::unordered_map` and a `std::vector`. The map is used to store object id (or indices) references to each of the objects while the `std::vector` stores all the objects in order (in accordance with the Table Construction for the [id24 dehacked spec](https://docs.google.com/document/d/19uWD7psvohLwQbzYl9Ja6q6SLCT5crN5HPFhUA9lcv8/edit?tab=t.0#heading=h.6jl50zhuy37m)). For future id24 spec updates, extending the container is straight forward.

### Handling Odamex Specific Objects

The id24 specification has allowance for source port specific objects using the index range `0xC0000000` to `0xFFFFFFFE`. This range is used for the odamex specific objects, most notably, those used in the Horde Multiplayer mode.

### Increasing The id24 Spec Compliance

The `mobj_t`, `state_t`, and `spriteinfo_t` structs now have their numerical index id in their struct in accordance with the [id24 spec structure specification](https://drive.google.com/drive/folders/1Q9GHoykGVJ7JONzYDXStM_vRdceL4Vi5).

Commit:
[WIP] dsdhacked

TODO:

- [X] Dynamically allocated sprite names
- [X] Dynamically allocated mobjinfo
- [X] Dynamically allocated state_t
- [ ] Odamex states appended to the end of the dynamically allocated
  states
- [X] Zone memory allocation for char* sprite names
- [ ] Zone memory allocation for state_t
- [ ] Zone memory allocation for mobjinfo objects (???)
- [ ] Dehacked Undo with dynamically allocated doom object collections
- [ ] Update use of NUMSPRITES where applicable
- [ ] Update use of NUMSTATES where applicable
- [ ] Update use of NUMMOBJTYPES where applicable

* Added dynamically allocated array for mobjinfo types
  * Includes new global variable num_mobjinfo_types and
    D_Initialize_mobjinfo function
  * These are managed by new functions defined in
    mobjinfo.h/mobjinfo.cpp
  * mobjinfo is initialized pointing at doom_mobjinfo. This saves an
    allocation and is safer when realloc-ing and replacing.
* Added dynamically allocated array for states
  * Includes a new global variable num_state_t_types and
    D_Initialize_states function
  * These are managed by new functions defined in state.h/state.cpp
  * states is not initialized from boomstates and odastates
* Added dynamically allocated array for sprite names
  * Includes new global variable num_spritet_types and
    D_Initialize_sprnames function
  * These are managed by new functions defined in sprite.h/sprite.cpp
  * sprnames is initialized pointing to doom_sprnames. This saves an
    allocation and is safer when realloc'ing and replacing.
* Base mobinfo types max is previous limit
* Base sprite types max is previous limit
* Base state types max is previous limit

Commit:
[WIP] dsdhacked

TODO:

- [X] Dynamically allocated sprite names
- [X] Dynamically allocated mobjinfo
- [X] Dynamically allocated state_t
- [ ] Odamex states appended to the end of the dynamically allocated
  states
- [X] Zone memory allocation for char* sprite names
- [ ] Zone memory allocation for state_t
- [ ] Zone memory allocation for mobjinfo objects (???)
- [ ] Dehacked Undo with dynamically allocated doom object collections
- [ ] Update use of NUMSPRITES where applicable
- [ ] Update use of NUMSTATES where applicable
- [ ] Update use of NUMMOBJTYPES where applicable

* Added dynamically allocated array for mobjinfo types
  * Includes new global variable num_mobjinfo_types and
    D_Initialize_mobjinfo function
  * These are managed by new functions defined in
    mobjinfo.h/mobjinfo.cpp
  * mobjinfo is initialized pointing at doom_mobjinfo. This saves an
    allocation and is safer when realloc-ing and replacing.
* Added dynamically allocated array for states
  * Includes a new global variable num_state_t_types and
    D_Initialize_states function
  * These are managed by new functions defined in state.h/state.cpp
  * states is not initialized from boomstates and odastates
* Added dynamically allocated array for sprite names
  * Includes new global variable num_spritet_types and
    D_Initialize_sprnames function
  * These are managed by new functions defined in sprite.h/sprite.cpp
  * sprnames is initialized pointing to doom_sprnames. This saves an
    allocation and is safer when realloc'ing and replacing.
* Base mobinfo types max is previous limit
* Base sprite types max is previous limit
* Base state types max is previous limit

Commit:
[WIP] dsdhacked

TODO:

- [X] Dynamically allocated sprite names
- [X] Dynamically allocated mobjinfo
- [X] Dynamically allocated state_t
- [ ] Odamex states appended to the end of the dynamically allocated
  states
- [X] Zone memory allocation for char* sprite names
- [ ] Zone memory allocation for state_t
- [ ] Zone memory allocation for mobjinfo objects (???)
- [ ] Dehacked Undo with dynamically allocated doom object collections
- [ ] Update use of NUMSPRITES where applicable
- [ ] Update use of NUMSTATES where applicable
- [ ] Update use of NUMMOBJTYPES where applicable

* Added dynamically allocated array for mobjinfo types
  * Includes new global variable num_mobjinfo_types and
    D_Initialize_mobjinfo function
  * These are managed by new functions defined in
    mobjinfo.h/mobjinfo.cpp
  * mobjinfo is initialized pointing at doom_mobjinfo. This saves an
    allocation and is safer when realloc-ing and replacing.
* Added dynamically allocated array for states
  * Includes a new global variable num_state_t_types and
    D_Initialize_states function
  * These are managed by new functions defined in state.h/state.cpp
  * states is not initialized from boomstates and odastates
* Added dynamically allocated array for sprite names
  * Includes new global variable num_spritet_types and
    D_Initialize_sprnames function
  * These are managed by new functions defined in sprite.h/sprite.cpp
  * sprnames is initialized pointing to doom_sprnames. This saves an
    allocation and is safer when realloc'ing and replacing.
* Base mobinfo types max is previous limit
* Base sprite types max is previous limit
* Base state types max is previous limit

Commit:
[WIP] dsdhacked

TODO:

- [X] Dynamically allocated sprite names
- [X] Dynamically allocated mobjinfo
- [X] Dynamically allocated state_t
- [ ] Odamex states appended to the end of the dynamically allocated
  states
- [X] Zone memory allocation for char* sprite names
- [ ] Zone memory allocation for state_t
- [ ] Zone memory allocation for mobjinfo objects (???)
- [ ] Dehacked Undo with dynamically allocated doom object collections
- [ ] Update use of NUMSPRITES where applicable
- [ ] Update use of NUMSTATES where applicable
- [ ] Update use of NUMMOBJTYPES where applicable

* Added dynamically allocated array for mobjinfo types
  * Includes new global variable num_mobjinfo_types and
    D_Initialize_mobjinfo function
  * These are managed by new functions defined in
    mobjinfo.h/mobjinfo.cpp
  * mobjinfo is initialized pointing at doom_mobjinfo. This saves an
    allocation and is safer when realloc-ing and replacing.
* Added dynamically allocated array for states
  * Includes a new global variable num_state_t_types and
    D_Initialize_states function
  * These are managed by new functions defined in state.h/state.cpp
  * states is not initialized from boomstates and odastates
* Added dynamically allocated array for sprite names
  * Includes new global variable num_spritet_types and
    D_Initialize_sprnames function
  * These are managed by new functions defined in sprite.h/sprite.cpp
  * sprnames is initialized pointing to doom_sprnames. This saves an
    allocation and is safer when realloc'ing and replacing.
* Base mobinfo types max is previous limit
* Base sprite types max is previous limit
* Base state types max is previous limit

Commit:
Fix issue with number of states in global state_t array being too small when loading objects in a level.

Commit:
Fix issue with number of states in global state_t array being too small when loading objects in a level.

Commit:
Add missing header and implementation for state_t handling.

Commit:
Add missing header and implementation for state_t handling.

cmbernard333 committed:
* Fixed issue in dehacked where a new mobjinfo wasn't having the correct doomednum set.
* Fixed issue in dehacked where debug print for sprites was causing index out of bounds when subspr wasn't set.
* Fixed issue in PatchCodePtrs where it was using the wrong max for number of states
* Added memset in mobjinfo and state reset functions to set to 0 ad there are parts of the code that depend on parts of either structure being 0.
* Use strncmp in sprite.cpp since windows doesn't have strncasecmp at that point.

cmbernard333 committed:
* Fixed issue in dehacked where a new mobjinfo wasn't having the correct doomednum set.
* Fixed issue in dehacked where debug print for sprites was causing index out of bounds when subspr wasn't set.
* Fixed issue in PatchCodePtrs where it was using the wrong max for number of states
* Added memset in mobjinfo and state reset functions to set to 0 ad there are parts of the code that depend on parts of either structure being 0.
* Use strncmp in sprite.cpp since windows doesn't have strncasecmp at that point.

cmbernard333 committed:
[WIP] Sprite replacement debugging.

cmbernard333 committed:
[WIP] Sprite replacement debugging.

cmbernard333 committed:
[WIP] Sprite replacement debugging.

cmbernard333 committed:
[WIP] Sprite replacement debugging.

cmbernard333 committed:
[WIP] Sprite replacement debugging.

cmbernard333 committed:
[WIP] Sprite replacement debugging.

cmbernard333 committed:
[WIP] Sprite replacement debugging.

cmbernard333 committed:
[WIP] Sprite replacement debugging.

cmbernard333 committed:
Fix sprnames not being correctly populated due to Z_Strdup not behaving as expected. Fix backups using wrong sizes in dehacked.

cmbernard333 committed:
Fix sprnames not being correctly populated due to Z_Strdup not behaving as expected. Fix backups using wrong sizes in dehacked.

Commit:
Change instances of NUMSPRITES, NUMMOBJTYPES, and NUMSTATES where applicable.

Commit:
Change instances of NUMSPRITES, NUMMOBJTYPES, and NUMSTATES where applicable.

Commit:
Change instances of NUMSPRITES, NUMMOBJTYPES, and NUMSTATES where applicable.

Commit:
Change instances of NUMSPRITES, NUMMOBJTYPES, and NUMSTATES where applicable.

cmbernard333 committed:
Fix base assignment of states and mobjinfo.

cmbernard333 committed:
Fix base assignment of states and mobjinfo.

cmbernard333 committed:
Fixed bug where new mobjinfo (specifically Things) would not render because the default value for translucency was being set to 0 when zeroing out the structs upon creation. Setting this to 0x1000 makes things appear but introduces a bug where particle containers now render with the IMP sprite.

cmbernard333 committed:
Fixed bug where new mobjinfo (specifically Things) would not render because the default value for translucency was being set to 0 when zeroing out the structs upon creation. Setting this to 0x1000 makes things appear but introduces a bug where particle containers now render with the IMP sprite.

cmbernard333 committed:
Correctly set sane defaults for state_t when re-alloc'ing. Only re-alloc minimum times needed for mobjinfo when increasing size.

cmbernard333 committed:
Correctly set sane defaults for state_t when re-alloc'ing. Only re-alloc minimum times needed for mobjinfo when increasing size.

cmbernard333 committed:
Re-add implementation of A_LineEffect.

cmbernard333 committed:
Re-add implementation of A_LineEffect.

Commit:
Remove un-needed std:: namespace specifier.

Commit:
Remove un-needed std:: namespace specifier.

cmbernard333 committed:
Fix issue with odamex wad command causing crash on startup. Tie all backup data structures for dsdhacked into a struct and remove static lifetime as it was causing issues on re-initialization during doom reboot wad. Add code to free doom object data structures if initialize is called again.

cmbernard333 committed:
Fix issue with odamex wad command causing crash on startup. Tie all backup data structures for dsdhacked into a struct and remove static lifetime as it was causing issues on re-initialization during doom reboot wad. Add code to free doom object data structures if initialize is called again.

Commit:
WIP odamex doom objects in separate file. Odamex global indices are managed in info.h/cpp but the objects themselves are in separate files.

Commit:
WIP odamex doom objects in separate file. Odamex global indices are managed in info.h/cpp but the objects themselves are in separate files.

Commit:
WIP odamex doom objects in separate file. Odamex global indices are managed in info.h/cpp but the objects themselves are in separate files.

Commit:
WIP odamex doom objects in separate file. Odamex global indices are managed in info.h/cpp but the objects themselves are in separate files.

Commit:
WIP odamex doom objects in separate file. Odamex global indices are managed in info.h/cpp but the objects themselves are in separate files.

Commit:
WIP odamex doom objects in separate file. Odamex global indices are managed in info.h/cpp but the objects themselves are in separate files.

cmbernard333 committed:
Add DoomObjectContainer<ObjType,Idx> as a replacement for the global doom object pointers states, sprnames, mobjinfo. Work-in-progress for negative indices as these data structures support mapping the index of a stored value to an IdxType (i.e. spritenum_t). Replaced the global sizes for the doom object pointers with functions that pass through to the new DoomObjectContainer instances.

cmbernard333 committed:
Add DoomObjectContainer<ObjType,Idx> as a replacement for the global doom object pointers states, sprnames, mobjinfo. Work-in-progress for negative indices as these data structures support mapping the index of a stored value to an IdxType (i.e. spritenum_t). Replaced the global sizes for the doom object pointers with functions that pass through to the new DoomObjectContainer instances.

cmbernard333 committed:
Add DoomObjectContainer<ObjType,Idx> as a replacement for the global doom object pointers states, sprnames, mobjinfo. Work-in-progress for negative indices as these data structures support mapping the index of a stored value to an IdxType (i.e. spritenum_t). Replaced the global sizes for the doom object pointers with functions that pass through to the new DoomObjectContainer instances.

cmbernard333 committed:
Add DoomObjectContainer<ObjType,Idx> as a replacement for the global doom object pointers states, sprnames, mobjinfo. Work-in-progress for negative indices as these data structures support mapping the index of a stored value to an IdxType (i.e. spritenum_t). Replaced the global sizes for the doom object pointers with functions that pass through to the new DoomObjectContainer instances.

cmbernard333 committed:
Add DoomObjectContainer<ObjType,Idx> as a replacement for the global doom object pointers states, sprnames, mobjinfo. Work-in-progress for negative indices as these data structures support mapping the index of a stored value to an IdxType (i.e. spritenum_t). Replaced the global sizes for the doom object pointers with functions that pass through to the new DoomObjectContainer instances.

cmbernard333 committed:
Add DoomObjectContainer<ObjType,Idx> as a replacement for the global doom object pointers states, sprnames, mobjinfo. Work-in-progress for negative indices as these data structures support mapping the index of a stored value to an IdxType (i.e. spritenum_t). Replaced the global sizes for the doom object pointers with functions that pass through to the new DoomObjectContainer instances.

cmbernard333 committed:
WIP doom_obj_container has more straight forward reset function.

cmbernard333 committed:
WIP doom_obj_container has more straight forward reset function.

Commit:
Fix dehacked backup doing a double free after converting backupsprnames to doom_obj_container.

Commit:
Fix dehacked backup doing a double free after converting backupsprnames to doom_obj_container.

cmbernard333 committed:
Just use vector inside of doom_obj_container. C++11 add .data() to it.

cmbernard333 committed:
Just use vector inside of doom_obj_container. C++11 add .data() to it.

cmbernard333 committed:
Initiailization logic is now uniform when starting client or server for all global doom object containers. Initial indexing insertion for doom objects is working inside of doom object container.

cmbernard333 committed:
Initiailization logic is now uniform when starting client or server for all global doom object containers. Initial indexing insertion for doom objects is working inside of doom object container.

Commit:
Move more code out of dehextra frame initialization in d_main and into the appropriate doom object functions. Add index lookup via the associative map inside of DoomObjectContainer. Add find function for DoomObjectContainer that returns NULL if it can't find an object. Started the transition of odamex specific states to negative indices.

Commit:
Move more code out of dehextra frame initialization in d_main and into the appropriate doom object functions. Add index lookup via the associative map inside of DoomObjectContainer. Add find function for DoomObjectContainer that returns NULL if it can't find an object. Started the transition of odamex specific states to negative indices.

cmbernard333 committed:
Removed ensure capacity functions for mobjinfo, states, and sprnames because DoomObjectContainer handles this now.

cmbernard333 committed:
Removed ensure capacity functions for mobjinfo, states, and sprnames because DoomObjectContainer handles this now.

cmbernard333 committed:
Can't look up frame against size as size doesn't correspond to position anymore.

cmbernard333 committed:
Can't look up frame against size as size doesn't correspond to position anymore.

Commit:
Added mobinfo command for printing mobjinfo entries. Updated more places using num_*_types functions to check index ranges. Fixed chainsaw and fist range.

Commit:
Added mobinfo command for printing mobjinfo entries. Updated more places using num_*_types functions to check index ranges. Fixed chainsaw and fist range.

cmbernard333 committed:
WIP odamex Things, Sprites, and States processing. Currently the side effect of operator[] creating entries in the DoomObjectContainer is problematic. It blows up in loops.

cmbernard333 committed:
WIP odamex Things, Sprites, and States processing. Currently the side effect of operator[] creating entries in the DoomObjectContainer is problematic. It blows up in loops.

cmbernard333 committed:
Dsdhacked using partial id24spec
---
Doom objects: mobjinfo (things), state_t(states), and sprnames(sprites)
are now all stored in DoomObjectContainer<ObjType,IdxType>. All types
now use int32_t (per id24 spec) to be referenced. The
DoomObjectContainer uses an OHashTable and a Vector internally to store
and address all objects.

All places in the code that previously used mobjinfo, states, and
sprnames now use the DoomObjectContainer and use pointers in place of
referencing the structs directly.

cmbernard333 committed:
Dsdhacked using partial id24spec
---
Doom objects: mobjinfo (things), state_t(states), and sprnames(sprites)
are now all stored in DoomObjectContainer<ObjType,IdxType>. All types
now use int32_t (per id24 spec) to be referenced. The
DoomObjectContainer uses an OHashTable and a Vector internally to store
and address all objects.

All places in the code that previously used mobjinfo, states, and
sprnames now use the DoomObjectContainer and use pointers in place of
referencing the structs directly.

cmbernard333 committed:
Move odamex objects to odamex_objects.cpp

cmbernard333 committed:
Move odamex objects to odamex_objects.cpp

cmbernard333 committed:
id24spec continued
------
* actor.h - sprite is now int32_t to match the value stored in spriteinfo_t
* cl_main.cpp - P_SetpSprite uses int32_t instead of spritenum_t
* d_dehacked.cpp - use int32_t for indices where appropriate. Sprite searching now does a search of sprnames instead of using org sprite logic.
* doom_obj_container.h - experimental template allowing InOrderContainer to be specified.
* info.h - added spriteinfo_t from id24 spec
* mobjinfo.cpp - remove span
* p_local.h - P_SetMobjState now uses int32_t instead of statenum_t
* p_mobj.cpp - static cast to statenum_t to check S_NULL for proper state shutdown
* p_mobj.h - P_setMobjState now uses int32_t instead of statenum_t
* p_pspr.cpp - P_SetPspriteptr uses int32_t instead of statenum_t
* p_setup.cpp - initializing sprites is now done with a sorted vector instead of a const char** namelist and a size. Sprite definitions are stored in a std::map instead of a pointer as sprite lookup is done via index and indices can be negative.
* r_data.cpp - precaching sprites is done using an unordered_map instead of the hitlist as indices can be negative.
* r_sprites.cpp - sprites is now a std::map and not a pointer as indices can be negative. Sprite reference by number uses int32_t instead of int.
* r_sprites.h - R_InitSprites uses a sorted vector instead of a namelist and count.
* r_state.h - sprites is a std::map instead of a pointer as indices can be negative.
* z_zone.cpp - fixed Z_StrDup.

cmbernard333 committed:
id24spec continued
------
* actor.h - sprite is now int32_t to match the value stored in spriteinfo_t
* cl_main.cpp - P_SetpSprite uses int32_t instead of spritenum_t
* d_dehacked.cpp - use int32_t for indices where appropriate. Sprite searching now does a search of sprnames instead of using org sprite logic.
* doom_obj_container.h - experimental template allowing InOrderContainer to be specified.
* info.h - added spriteinfo_t from id24 spec
* mobjinfo.cpp - remove span
* p_local.h - P_SetMobjState now uses int32_t instead of statenum_t
* p_mobj.cpp - static cast to statenum_t to check S_NULL for proper state shutdown
* p_mobj.h - P_setMobjState now uses int32_t instead of statenum_t
* p_pspr.cpp - P_SetPspriteptr uses int32_t instead of statenum_t
* p_setup.cpp - initializing sprites is now done with a sorted vector instead of a const char** namelist and a size. Sprite definitions are stored in a std::map instead of a pointer as sprite lookup is done via index and indices can be negative.
* r_data.cpp - precaching sprites is done using an unordered_map instead of the hitlist as indices can be negative.
* r_sprites.cpp - sprites is now a std::map and not a pointer as indices can be negative. Sprite reference by number uses int32_t instead of int.
* r_sprites.h - R_InitSprites uses a sorted vector instead of a namelist and count.
* r_state.h - sprites is a std::map instead of a pointer as indices can be negative.
* z_zone.cpp - fixed Z_StrDup.

cmbernard333 committed:
id24spec/horde fixed continued
------
* doom_obj_container.h - added contains function
* odamex_objects.cpp - added correct doomednum values for MT_HORDESPAWN and MT_CAREPACKAGE
* p_mobj.cpp - simplified looking up mobjinfo in spawn_map
* r_things.cpp - fixed issued where negative indices for sprites were not being found. Added better logging for when a sprite or sprite frame isn't found.

cmbernard333 committed:
id24spec/horde fixed continued
------
* doom_obj_container.h - added contains function
* odamex_objects.cpp - added correct doomednum values for MT_HORDESPAWN and MT_CAREPACKAGE
* p_mobj.cpp - simplified looking up mobjinfo in spawn_map
* r_things.cpp - fixed issued where negative indices for sprites were not being found. Added better logging for when a sprite or sprite frame isn't found.

cmbernard333 committed:
* clean up some todos

cmbernard333 committed:
* clean up some todos

cmbernard333 committed:
* sv_level.cpp - fixed pointer syntax
* clean up some todos

cmbernard333 committed:
* sv_level.cpp - fixed pointer syntax
* clean up some todos

cmbernard333 committed:
* fix release build issue
* update backup code for dehacked

cmbernard333 committed:
* fix release build issue
* update backup code for dehacked

cmbernard333 committed:
* fix consecutive right angle bracket error.

cmbernard333 committed:
* fix consecutive right angle bracket error.

rakohus commented:
Bug encountered in eviternity map35: 
![image](https://github.com/user-attachments/assets/07384b23-809a-4e87-b3b5-621fefb010e9)

rakohus commented:
Bug encountered in id1.wad map01. If you blow up the blue lamps in the first room that the zombiemen are in, it plays the odamex team chat sound on their "death".

rakohus commented:
Bug: sd21.wad doesn't load. https://www.doomworld.com/forum/topic/133318-smooth-doom-on-mbf21-dsdhacked-rifle-addon-and-standard-v61-released-04oct2024/

Commit:
Fix c++11 compatibility with gcc.

Commit:
Fix c++11 compatibility with gcc.

cmbernard333 committed:
Update operator<< for FArchive for state_t and mobjinfo_t.

cmbernard333 committed:
Update operator<< for FArchive for state_t and mobjinfo_t.

cmbernard333 committed:
Replace auto with iterator declaration.

cmbernard333 committed:
Replace auto with iterator declaration.

cmbernard333 committed:
Use mobjinfo type instead of calculating the offset with pointers.

cmbernard333 committed:
Use mobjinfo type instead of calculating the offset with pointers.

cmbernard333 committed:
More C++11 linux incompatibility fixes.

cmbernard333 committed:
More C++11 linux incompatibility fixes.

cmbernard333 committed:
Fixed fmt format issue by using statenum instead of pointer difference.

cmbernard333 committed:
Fixed fmt format issue by using statenum instead of pointer difference.

cmbernard333 committed:
Fix explicit copy constructor.

cmbernard333 committed:
Fix explicit copy constructor.

cmbernard333 committed:
Fix I_MSTime forward decl.

cmbernard333 committed:
Fix I_MSTime forward decl.

cmbernard333 committed:
Fixed logic error when loading initial mobjinfo where a separate idx was being used to load the mobjinfo lookup table. This was causing as strange bug during tics where certain dehacked projectiles were having their subsector set to NULL.

cmbernard333 committed:
Fixed logic error when loading initial mobjinfo where a separate idx was being used to load the mobjinfo lookup table. This was causing as strange bug during tics where certain dehacked projectiles were having their subsector set to NULL.

cmbernard333 committed:
Account for the NULL state when a player joins the server.

cmbernard333 committed:
Account for the NULL state when a player joins the server.

cmbernard333 committed:
Fix PatchText issue by using iterator instead of iterating by index.

cmbernard333 committed:
Fix PatchText issue by using iterator instead of iterating by index.

electricbrass committed:
Implement PatchSounds for DSDHacked Bex-style [SOUNDS]

electricbrass committed:
Add SoundMap reinitialization to D_UndoDehPatch

electricbrass committed:
Don't use string table for DSDHacked sound replacements

electricbrass committed:
Add SoundMap backup to BackupData

electricbrass committed:
Don't call SoundMap constructor in header file

electricbrass committed:
Use iterators instead of comparisons against SoundMap size

electricbrass committed:
Don't search SoundMap twice

electricbrass committed:
Allow PatchSounds to work with only a single dehacked pass

electricbrass committed:
Free sound index strings made with strdup

electricbrass committed:
Store mobjinfo and state indices to archives as int32_t instead of WORD

electricbrass committed:
Update wxwidgets package to new name from ubuntu 24.04

cmbernard333 committed:
Merge branch 'feature/dsdhacked-2021' of github.com:cmbernard333/odamex into feature/dsdhacked-2021

cmbernard333 committed:
Merge pull request #2 from electricbrass/dsdhacked-warmup-fix

[BUGFIX] Error at end of warmup timer - Linux build issue missing PNG noted. Can re-check when merging to `stable`.

cmbernard333 committed:
Merge branch 'feature/dsdhacked-2021' of github.com:cmbernard333/odamex into feature/dsdhacked-2021

cmbernard333 committed:
Fix issue where size 0 sprite lumps were causing an out of bounds error when checking MAX_SPRITE_FRAMES when installing sprites.

electricbrass committed:
Use const char* instead of char** in IdxToSoundName

electricbrass committed:
Check for empty string in IdxToSoundName and don't set default ripsound to empty string

electricbrass committed:
Less manual memory management when patching [SOUNDS]

electricbrass committed:
Strip newlines from sound names

electricbrass committed:
Initialize SoundMap on the server

cmbernard333 committed:
Merge pull request #1 from electricbrass/dsdhacked-sound

DSDHacked [SOUNDS] support

rakohus committed:
Merge branch 'stable' into feature/dsdhacked-2021

electricbrass committed:
Reinitialize states, sprites, and mobjinfo from full contents of backup

Also changes a number of instances of iteration over DoomObjContainers to use iterators

electricbrass committed:
Call S_ClearSoundLumps in D_Shutdown instead of in S_ParseSndInfo

electricbrass committed:
Clean up some compiler warnings

electricbrass committed:
Update D_PostProcessDeh

electricbrass committed:
More cleanup of usage of DoomObjectContainers, changed LookupTable type to std::unordered_map

electricbrass committed:
Revert DoomObjectContainer to using OHashTable, clear with clear because it isn't buggy like erase and is more efficient

electricbrass committed:
Don't invalide iterators in OHashTable::erase(iterator it1, iterator it2)

cmbernard333 committed:
Merge pull request #3 from electricbrass/wad-switch-fix

Fix for `wad` command

cmbernard333 committed:
Cleanup doomobjcontainer template definition by removing InOrderContainer (just default to vector internally). Added FreeFunction template parameter used for freeing objects internally on clear and object destruction. FreeFunction defaults to noop lambda.

bcahue committed:
merge up

cmbernard333 committed:
Merge branch 'feature/dsdhacked-2021' of github.com:cmbernard333/odamex into feature/dsdhacked-2021

cmbernard333 committed:
Move Z_Init() out of D_Init() and call it before calling dehacked to allow using it during patch processing.

cmbernard333 committed:
- Move Z_Init() to before the initialization of doom vanilla objs (Mobjinfo, States, Sprite names).
- Use Z_Malloc for initialization of doom vanilla objs
- Remove specific clear() specialized template function for doom obj container storing const char*

cmbernard333 committed:
Fix issue where index for mobjinfo and sprites was being compared to size when it should be using .find().

cmbernard333 committed:
Fix too many arguments to Z_StrDup.

cmbernard333 committed:
- Move the initialization of doom objects to a common function in common/d_main.cpp that is called at start up for the client and server. This is also called by the `wad` command during reboot.

cmbernard333 committed:
Use Z_Malloc for odamex objects.

cmbernard333 committed:
Merge remote-tracking branch 'origin/protobreak' into feature/dsdhacked-2021

bcahue committed:
remove berserk puffs

bcahue committed:
remove setting this when playing demos (since it doesn't desync demos anymore)

bcahue committed:
only show berserk effect when fist is out

cmbernard333 committed:
Fix mobjinfo struct member reference.

cmbernard333 committed:
Merge remote-tracking branch 'origin/protobreak' into feature/dsdhacked-2021

cmbernard333 committed:
Fix reference to caststate.

electricbrass committed:
Don't call strdup on null pointer when backing up sounds

cmbernard333 committed:
Remove the dereference on m_Source when casting to (DObject*&) when serializing a pusher. Add additional error cases checked in GetNumForName.

cmbernard333 committed:
Merge pull request #4 from electricbrass/dsdhacked

Fix crash on non-Windows systems

electricbrass committed:
Bump SDL Mixer to 2.6.3

electricbrass committed:
Fix cmake and lower version to 2.6.2

electricbrass committed:
Fix incorrect checksum

electricbrass committed:
Fix windows x32 artifact uploads

cmbernard333 committed:
- Fix issue with callvote restart on maps with specific pusher types.
  - Pushers were being unlinked causing a null ptr.
- Demo processing now uses OLumpName instead of const char*

cmbernard333 committed:
Merge remote-tracking branch 'origin/protobreak' into feature/dsdhacked-2021

bcahue committed:
fix sdl version on release scripts

bcahue committed:
find the right dlls to copy

bcahue committed:
fix setup installer as well

bcahue committed:
fix inno setup

cmbernard333 committed:
Merge remote-tracking branch 'origin/protobreak' into feature/dsdhacked-2021

cmbernard333 committed:
Fix issue with dehacked weapons using codeptrs where they wouldn't fire or make a sound.

electricbrass committed:
Address MELEERANGE redefinition warning

electricbrass committed:
Merge branch 'protobreak' into dsdhacked-merge

electricbrass committed:
Move <map> into odamex.h

electricbrass committed:
Address some unreferenced local variable warnings

cmbernard333 committed:
New fix for calculating the correct psprnum for the action being called. This one works with eviternity 2 and doosk.

electricbrass committed:
Merge remote-tracking branch 'cmbernard333/feature/dsdhacked-2021' into dsdhacked-merge

electricbrass committed:
Merge branch 'protobreak' into dsdhacked-merge

electricbrass committed:
Merge branch 'protobreak' into dsdhacked-merge

electricbrass committed:
Merge branch 'protobreak' into dsdhacked-merge

cmbernard333 committed:
Merge pull request #5 from electricbrass/dsdhacked-merge

Merge `protobreak` into `feature/dsdhacked-2021`

cmbernard333 committed:
feat: dsdhacked

Summary

* Fix the WAD command crashing when looking up if a sprite has no frames when loading a thing in a map

Implementation

* Dehacked now allocates backup objects to the heap so they are not cleared when the zone memory is cleared. During D_Shutdown(), when the DEH patch is undone, it restores from the heap into zone memory.
* Doom Object container no longer uses OHashTable and instead uses std::unordered_map
* Fixed a bug where the spawn_map wasn't cleared on reinitializing during a WAD reboot and initial load

Investigations

* Heap thrasing on wad command change. DoomWadReboot (i.e. WAD cmd) undoes the DEH patch then clears the zone memory then reinitializes all doom objects. Ideally, we would just reinitialize the doom state to expected, which is either previous or clean slate during dehacked undo (i.e. D_UndoDehPatch), but
currently the code does not distinguish between the possible states of shutdown, reboot, and reloading after resources fail to load.

cmbernard333 committed:
Fix issue with sound not playing for certain weapons.

cmbernard333 committed:
Remove some todos.
